src/proxy/ssh/server.ts

```diff
@@ -0,0 +1,1170 @@
+      clearTimeout(captureTimeout);
+      await originalEndHandler();
+    };
+
+    const timeoutAwareErrorHandler = (error: Error) => {
+      clearTimeout(captureTimeout);
+      errorHandler(error);
+    };
+
+    // Attach event handlers
+    stream.on('data', dataHandler);
+    stream.once('end', timeoutAwareEndHandler);
+    stream.on('error', timeoutAwareErrorHandler);
+  }
+
+  private async handlePullOperation(
+    command: string,
+    stream: ssh2.ServerChannel,
+    client: ClientWithUser,
+    repoPath: string,
+    gitPath: string,
+  ): Promise<void> {
+    console.log(`[SSH] Handling pull operation for ${repoPath}`);
+    const hostHeader = this.resolveHostHeader();
+
+    // For pull operations, execute chain first (no pack data to capture)
+    const req = {
+      originalUrl: `/${repoPath}/${gitPath}`,
+      url: `/${repoPath}/${gitPath}`,
+      method: 'GET' as const,
+      headers: {
+        'user-agent': 'git/ssh-proxy',
+        'content-type': 'application/x-git-upload-pack-request',
+        host: hostHeader,
+        'x-forwarded-proto': 'https',
+        'x-forwarded-host': hostHeader,
+      },
+      body: null,
+      user: client.authenticatedUser || null,
+      isSSH: true,
+      protocol: 'ssh' as const,
+      sshUser: {
+        username: client.authenticatedUser?.username || 'unknown',
+        email: client.authenticatedUser?.email,
+        gitAccount: client.authenticatedUser?.gitAccount,
+        sshKeyInfo: client.userPrivateKey,
+      },
+      authContext: this.buildAuthContext(client),
+    };
+
+    const res = {
+      headers: {},
+      statusCode: 200,
+      set: function (headers: any) {
+        Object.assign(this.headers, headers);
+        return this;
+      },
+      status: function (code: number) {
+        this.statusCode = code;
+        return this;
+      },
+      send: function (data: any) {
+        return this;
+      },
+    };
+
+    // Execute the proxy chain
+    try {
+      console.log(`[SSH] Executing security chain for pull operation`);
+      const result = await chain.executeChain(req, res);
+      if (result.error || result.blocked) {
+        const message =
+          result.errorMessage || result.blockedMessage || 'Request blocked by proxy chain';
+        throw new Error(message);
+      }
+
+      console.log(`[SSH] Security chain passed, connecting to remote`);
+      // Chain passed, connect to remote Git server
+      await this.connectToRemoteGitServer(command, stream, client);
+    } catch (chainError: unknown) {
+      console.error(
+        `[SSH] Chain execution failed for user ${client.authenticatedUser?.username}:`,
+        chainError,
+      );
+      const errorMessage = chainError instanceof Error ? chainError.message : String(chainError);
+      stream.stderr.write(`Access denied: ${errorMessage}\n`);
+      stream.exit(1);
+      stream.end();
+      return;
+    }
+  }
+
+  private async forwardPackDataToRemote(
```

Comment:
It seems this function is similar to `connectToRemoteGitServer`, maybe we could extract the connection logic into its own util function to reduce the number of lines?